### PR TITLE
kernels: add volk_32fc_x2_s32fc_multiply_conjugate_32fc

### PR DESCRIPTION
**Pull request by bstalk**
# Raspberry Pi 4B
./volk_32fc_x2_s32fc_multiply_conjugate_32fc_test.sh

RUN_VOLK_TESTS: volk_32fc_x2_s32fc_multiply_conjugate_32fc(131071,1)
neon completed in 1.08767 ms
generic completed in 3.07554 ms
Best aligned arch: neon
Best unaligned arch: neon

# Intel
./volk_32fc_x2_s32fc_multiply_conjugate_32fc_test.sh

RUN_VOLK_TESTS: volk_32fc_x2_s32fc_multiply_conjugate_32fc(131071,1)
a_avx completed in 0.280102 ms
a_sse3 completed in 0.286507 ms
generic completed in 1.88841 ms
Best aligned arch: a_avx
Best unaligned arch: generic

### CODE REVIEW

**Comment by ast** [comment]
Fantastic to see progress in ARM NEON!

**Comment by michaelld** [comment]
CI failure has nothing to do with this PR.

**Review by michaelld — DISMISSED** [review]
LGTM

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (jdemel) [inline kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: Could you move these lines to `include/volk_sse3_intrinsics.h`? Also, that would make it easier to add an unaligned version too. If you could add an unaligned version, that would be great.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (jdemel) [inline kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: Could you move these lines to `include/volk_avx_intrinsics.h`? Also, that would make it easier to add an unaligned version too. If you could add an unaligned version, that would be great.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (jdemel) [inline kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: I suggest to update the description here. Specifically be more verbose with the inputs.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (jdemel) [inline kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: A short example similar to the ones you can find in other kernels would be helpful.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (jdemel) [inline kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: I'm not sure but we mostly use `{}` for one-liners. It's something I prefer and recommend at least.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (jdemel) [inline kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: Can you move the generic kernel to the top? In order to understand a kernel, I always look at the generic impl first. It's helpful to have it as the first impl.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (jdemel) [inline kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: It should be sufficient to have one include guard for this file. They don't need to be separated for aligned/unaligned versions. Also, that would make those includes here obsolete.

**Review by jdemel — CHANGES_REQUESTED** [review]
This looks like a nice addition. Maybe an FMA version would make sense too.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (bstalk) [reply kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: Yes, I will move the generic kernel to the top of this file.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (bstalk) [reply kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: I will use {}.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (bstalk) [reply kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: Thank you for reviewing. I have a question.

"move these lines to include/volk_avx_intrinsics.h?" means:
1. "add method to include/volk_avx_intrinsics.h", for EXAMPLE:
static inline __m256 _mm256_complexconjugatemulscalar_ps(__m256 x, __m256 y, __m256 sl, __m256 sh) {
   __m256 z, tmp1, tmp2;
   __m256 conjugator = _mm256_setr_ps(0, -0.f, 0, -0.f, 0, -0.f, 0, -0.f);
   x = _mm256_xor_ps(x, conjugator);
   tmp1 = _mm256_mul_ps(x,sl);
   x = _mm256_shuffle_ps(x,x,0xB1);
   tmp2 = _mm256_mul_ps(x,sh);
   z = _mm256_addsub_ps(tmp1,tmp2);
   return _mm256_add_ps(y, z);
}

2. and "volk_fc32_x2_s32fc_multiply_conjugate_32fc_a_avx()?", for EXAMPLE:
...
for (;number < quarterPoints; number++) {
    x = _mm256_load_ps((float*)b);
    y = _mm256_load_ps((float*)a);
    z = _mm256_complexconjugatemulscalar_ps(x, y, sl, sh);
    _mm256_store_ps((float*)c,z);
...

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (bstalk) [reply kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: Thanks.
I try to add more verbose with the inputs.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (bstalk) [reply kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: There is a same question as avx's comment.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (bstalk) [reply kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: I will remove this guard, header will have one guard.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (jdemel) [reply kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: Yes. that's it. There is already some code for conjugation and multiplication. I assume you can at least use the conjugate function. Maybe you can reuse complex multiplication as well. Though, I suggest to not force your code to use this function. only if it is a one-to-one replacement like in the conjugator case.

**Review by jdemel — COMMENTED** [review]
Thanks for working on this PR.

File: kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h
Comment (bstalk) [reply kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h]: I might find this can be realized by combinations of existing intrinsics... I'm studying now.
Thanks for your suggestion.

**Comment by bstalk** [comment]
A lot of advice by @jj1bdx. Thanks.

**Review by michaelld — COMMENTED** [review]
Is this ready for a final review?

**Comment by bstalk** [comment]
Yes, would you please review this?

**Comment by michaelld** [comment]
> Yes, would you please review this?

Sure. Guessing I'll be able to get to it tomorrow.

**Comment by michaelld** [comment]
Works for me. @jdemel  Are all of the requests dealt with?

**Review by jdemel — APPROVED** [review]
LGTM now. Thanks for the PR.

**Comment by michaelld** [comment]
@bstalk do you have a CLA (code license agreement) setup with the GNU folks yet? Guessing you don't ... If not, please email @bhilburn : bhilburn@gnuradio.org ... we need one for this much code change.

**Review by michaelld — APPROVED** [review]
LGTM; thanks for the submission!

Just need the CLA to get this merged.

**Comment by rear1019** [comment]
Is there any reason why “add” is not part of the function’s name? The current name does not imply (explicitly) that the functions performs any other operation than multiplication. The only hint in the function’s name is the fact that three input arguments are expected – which requires knowledge of VOLK’s naming convention.

**Comment by michaelld** [comment]
> Is there any reason why “add” is not part of the function’s name? The current name does not imply (explicitly) that the functions performs any other operation than multiplication. The only hint in the function’s name is the fact that three input arguments are expected – which requires knowledge of VOLK’s naming convention.

Good catch! Now ... the next question is ... what's a better name?

**Comment by bstalk** [comment]
Thanks for indication.
How about "volk_32fc_x2_s32fc_multiply_conjugate_add_32fc" ?

**Comment by jdemel** [comment]
"volk_32fc_x2_s32fc_multiply_conjugate_add_32fc" would follow our naming scheme and express what the kernel really does. That'd be a good name imho.

**Comment by michaelld** [comment]
agreed! please make it so, @bstalk !

**Comment by bstalk** [comment]
I made it.
@bhilburn: By the way, name of new file was changed by this update. Need to re-submit CLA?

**Comment by bhilburn** [comment]
Hi @bstalk - No, the one you submitted covers all contributions to the
project.

I see the FSF sent you back the contract. As soon as you sign and send it
back, we're good for this and any other contribution you make :)

**Comment by bstalk** [comment]
@bhilburn, Thank you so much. I will sign and send back it as soon as possible.

By the way, I will push one more, I forgot to modify the function name between \code ... \endcode. Sorry to disturb.

**Comment by michaelld** [comment]
@bstalk @bhilburn please let us know when you have confirmation of CLA success with the FSF. I'd prefer to wait until then to merge.

**Comment by bstalk** [comment]
I sent signed and scanned form about 6 hours ago.

**Comment by bstalk** [comment]
I got the form which FSF signed.

**Comment by michaelld** [comment]
Excellent!

**Comment by michaelld** [comment]
Has CLA. CI passes. Multiple positive reviews. Merging.

**Pull request merged**
